packages/asset-swapper/src/utils/aggregate.ts

```diff
@@ -0,0 +1,800 @@
+            .queryOrdersAndSampleBuys(nativeOrders, signatures, sources.map(s => SOURCE_TO_ADDRESS[s]), sampleAmounts)
+            .callAsync();
+    } else {
+        [fillableAmount, rawSamples] = await samplerContract
+            .queryOrdersAndSampleSells(nativeOrders, signatures, sources.map(s => SOURCE_TO_ADDRESS[s]), sampleAmounts)
+            .callAsync();
+    }
+    const quotes = rawSamples.map((rawDexSamples, sourceIdx) => {
+        const source = sources[sourceIdx];
+        return rawDexSamples.map((sample, sampleIdx) => ({
+            source,
+            input: sampleAmounts[sampleIdx],
+            output: sample,
+        }));
+    });
+    return [fillableAmount, quotes];
+}
+
+/**
+ * Compute the total output for a fill path, optionally clipping the input
+ * to `maxInput`.
+ */
+export function getPathOutput(path: Fill[], maxInput?: BigNumber): BigNumber {
+    let currentInput = ZERO;
+    let currentOutput = ZERO;
+    for (const fill of path) {
+        if (maxInput && currentInput.plus(fill.input).gte(maxInput)) {
+            const partialInput = maxInput.minus(currentInput);
+            currentOutput = currentOutput.plus(getPartialFillOutput(fill, partialInput));
+            currentInput = partialInput;
+        } else {
+            currentInput = currentInput.plus(fill.input);
+            currentOutput = currentOutput.plus(fill.output);
+        }
+    }
+    return currentOutput;
+}
+
+/**
+ * Generate sample amounts up to `maxFillAmount`.
+ */
+export function getSampleAmounts(maxFillAmount: BigNumber, numSamples: number): BigNumber[] {
+    const amounts = [];
+    for (let i = 0; i < numSamples; i++) {
+        amounts.push(
+            maxFillAmount
+                .times(i + 1)
+                .div(numSamples)
+                .integerValue(BigNumber.ROUND_UP),
+        );
+    }
+    return amounts;
+}
+
+// Gets the difference between two sets.
+function difference<T>(a: T[], b: T[]): T[] {
+    return a.filter(x => b.indexOf(x) === -1);
+}
+
+function createSellPathFromNativeOrders(orders: SignedOrderWithoutDomain[], fillableAmounts: BigNumber[]): Fill[] {
+    const path: Fill[] = [];
+    // tslint:disable-next-line: prefer-for-of
+    for (let i = 0; i < orders.length; i++) {
+        const order = orders[i];
+        const takerAmount = fillableAmounts[i];
+        const makerAmount = getOrderFillableMakerAmount(order, takerAmount);
+        // Native orders can be filled in any order, so they're all root nodes.
+        path.push({
+            flags: FillFlags.SourceNative,
+            exclusionMask: 0,
+            input: takerAmount,
+            output: makerAmount,
+            data: {
+                source: ERC20BridgeSource.Native,
+                order,
+            },
+        });
+    }
+    return path;
+}
+
+function createBuyPathFromNativeOrders(orders: SignedOrderWithoutDomain[], fillableAmounts: BigNumber[]): Fill[] {
```

Comment:
Can we reuse some of the logic from above? (as in dedup the logic) Not a big issue

Reply:
Turns out these are subtly different enough to make it not worth it.